### PR TITLE
sigstore: ratchet down the bundle certs

### LINKED ISSUE #630: Bundle: check and generate the certificate chain more aggressively
Now that https://github.com/sigstore/protobuf-specs/issues/75 is resolved with stronger requirements around the certificate chain in Sigstore bundles, `sigstore-python` should perform stronger enforcement checks on the bundles that it both loads and generates.

In particular:

1. We should add unit tests that ensure that we never generate bundles containing intermediate or root CAs;
2. We should add unit tests that ensure that we emit a warning when loading a bundle that contains intermediate or root CAs;
3. We should add unit tests that ensure that we produce a hard failure when the first member of the bundle's certificate chain is **not** a non-CA leaf.

### PR DESCRIPTION
**Pull request by woodruffw**
~~WIP.~~

Closes #630.

### CODE REVIEW

**Comment by woodruffw** [comment]
This should be good to go. The diff is pretty large, but the majority of it is in the checked-in test assets (a whole bunch of slightly misshapen certificates). Many of those cases _should_ be caught by chain building and validation, but having them additionally here ensures that underlying changes don't bubble up unexpectedly as behavioral changes to what we consider well-formed certificates in the context of Sigstore.

File: sigstore/verify/models.py
Comment (woodruffw) [inline sigstore/verify/models.py:253]: Flagging for review: we only produce a warning here, not an error. This is because, despite the `MUST NOT` in the bundle spec, some bundles might already exist with root and (redundant) intermediate CAs in them.

File: sigstore/_utils.py
Comment (woodruffw) [reply sigstore/_utils.py:214]: Yep, good catch!

File: sigstore/verify/models.py
Comment (woodruffw) [reply sigstore/verify/models.py:253]: Yep, good idea.

File: sigstore/_utils.py
Comment (woodruffw) [reply sigstore/_utils.py:225]: `KeyUsage`'s criticality is actually a `SHOULD`, but this is perhaps somewhere that Sigstore as an ecosystem can be stronger.

cc @znewman01 @haydentherapper WDYT about a client side requirement that CA certs assert `KeyUsage.critical`? This shouldn't be a breaking change, since all existing CA certs should already have this extension marked correctly.

File: sigstore/_utils.py
Comment (woodruffw) [reply sigstore/_utils.py:214]: Done.

File: sigstore/verify/models.py
Comment (woodruffw) [reply sigstore/verify/models.py:272]: Thinking about this a bit more: it'd be nice to have a stable URL to point to here. 

Do the `protobuf-specs` docstrings get rendered anywhere? CC @znewman01 @haydentherapper

File: sigstore/verify/models.py
Comment (tetsuo-cpp) [inline sigstore/verify/models.py:258]: Nit: Maybe do the certificate loading ahead of time? Then the code below will only contain the checks rather than doing it in two different spots.
```python
leaf_certs, *chain_raw = [load_der_x509_certificate(cert.raw_bytes) for cert in certs]

# Now do checks...
```

File: test/unit/assets/x509/build-testcases.py
Comment (tetsuo-cpp) [inline test/unit/assets/x509/build-testcases.py:23]: This doesn't need to be done in this PR, but would it seem reasonable for this to run as part of the `Makefile` (and if it changes files, breaks CI?).

That'd ensure that this file doesn't get out of sync with the checked in certs and it keeps working with future `cryptography` versions.

**Review by tetsuo-cpp — DISMISSED** [review]
LGTM!

File: sigstore/verify/models.py
Comment (woodruffw) [reply sigstore/verify/models.py:258]: Good call!

File: test/unit/assets/x509/build-testcases.py
Comment (woodruffw) [reply test/unit/assets/x509/build-testcases.py:23]: Yeah, good idea -- I should probably make it use fixed keys and times as well, to ensure that it's reproducible.

File: sigstore/verify/models.py
Comment (tetsuo-cpp) [reply sigstore/verify/models.py]: [728f7f2](https://github.com/sigstore/sigstore-python/pull/632/commits/728f7f2c92b4c7b2d990f004163b28d45d07a6c2)

File: test/unit/assets/x509/build-testcases.py
Comment (woodruffw) [reply test/unit/assets/x509/build-testcases.py:23]: (Let's create a separate issue to track this, since it isn't a blocker.)

File: sigstore/_utils.py
Comment (woodruffw) [reply sigstore/_utils.py:307]: Same as https://github.com/sigstore/sigstore-python/pull/632#discussion_r1188996799 -- until Sigstore as an ecosystem establishes CA guidance here, we probably shouldn't assert this criticality bit.

File: test/unit/assets/x509/build-testcases.py
Comment (tetsuo-cpp) [reply test/unit/assets/x509/build-testcases.py:23]: I think I have this working already so I'll push it up. If you feel it needs more work, feel free to revert it back out.

Comment (tetsuo-cpp) [reply test/unit/assets/x509/build-testcases.py:23]: [151d4f0](https://github.com/sigstore/sigstore-python/pull/632/commits/151d4f034038a29056380b015352782ed4535f55)

**Review by tetsuo-cpp — APPROVED** [review]

File: test/unit/assets/x509/build-testcases.py
Comment (woodruffw) [reply test/unit/assets/x509/build-testcases.py:23]: Cool, will take a look shortly 🙂

File: .github/workflows/lint.yml
Comment (woodruffw) [inline .github/workflows/lint.yml:76]: Good call adding this!

File: test/unit/assets/x509/build-testcases.py
Comment (woodruffw) [reply test/unit/assets/x509/build-testcases.py:23]: Looks great, thank you for breaking the keys and other nondeterminism out!

**Comment by woodruffw** [comment]
Needs one last re-app 🙂

**Review by tetsuo-cpp — APPROVED** [review]
LGTM! Let's get this in.

**Comment by tetsuo-cpp** [comment]
Oops, still needs another approval since I made the last push.

**Comment by woodruffw** [comment]
Hmm, possibly a weird GitHub hiccup: I can't merge this because it thinks that some very ancient CI requirements (which should have been removed a long time ago) are unmet:

Screenshot 2023-05-12 at 10 10 22 AM

**Comment by cpanato** [comment]
fixed here https://github.com/sigstore/community/pull/271

**Review by tetsuo-cpp — APPROVED** [review]

**Review by tetsuo-cpp — APPROVED** [review]

**Pull request merged**
